numpy/core/src/multiarray/array_method.c

```diff
@@ -163,6 +166,55 @@ npy_default_get_strided_loop(
 }
 
 
+/* Declared in `ufunc_object.c` */
+NPY_NO_EXPORT PyObject *
+PyUFunc_GetIdentity(PyUFuncObject *ufunc, npy_bool *reorderable);
+
+/*
+ * The default `get_identity` attempts to look up the identity from the
+ * calling ufunc.
+ */
+static int
+default_get_identity_function(PyArrayMethod_Context *context,
+        char *item, NPY_ARRAYMETHOD_IDENTITY_FLAGS *flags)
+{
+    *flags = 0;
+
+    PyUFuncObject *ufunc = (PyUFuncObject *)context->caller;
+    if (!PyObject_TypeCheck(ufunc, &PyUFunc_Type)) {
+        /* Should not happen, but if it does, just give up */
+        return 0;
+    }
+
+    npy_bool reorderable;
+    PyObject *identity_obj = PyUFunc_GetIdentity(ufunc, &reorderable);
+    if (identity_obj == NULL) {
+        return -1;
+    }
+    if (reorderable) {
+        *flags |= NPY_METH_IS_REORDERABLE;
+    }
+    if (item == NULL) {
+        /* Only reorderable flag was requested (user provided initial value) */
+        Py_DECREF(identity_obj);
+        return 0;
+    }
+
+    if (identity_obj != Py_None) {
+        /* We do not consider the value an identity for object dtype */
+        *flags |= NPY_METH_ITEM_IS_DEFAULT;
+        if (context->descriptors[0]->type_num != NPY_OBJECT) {
+            *flags |= NPY_METH_ITEM_IS_IDENTITY;
+        }
+        int res = PyArray_Pack(context->descriptors[0], item, identity_obj);
```

Comment:
I wonder if it is really the best to do this packing inside this function, instead of trying to interpret the result. See main comment.

Reply:
I have a slight preference for using the raw data, also because I am still wondering if we might want dtypes without any scalar at all, and creating an array just to get the identity seems a bit heavy.

Alternatively, I could actually also `Pack` the `initial` provided by the user and use the `raw` version for both, although this version has the nice property that allowing arrays is changing one line to `CopyObject` :).